flink-tests/src/test/java/org/apache/flink/api/connector/source/lib/NumberSequenceSourceITCase.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.connector.source.lib;
+
+import org.apache.flink.api.common.restartstrategy.RestartStrategies;
+import org.apache.flink.streaming.api.CheckpointingMode;
+import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.util.StreamCollector;
+
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.LongStream;
+
+import static org.junit.Assert.assertArrayEquals;
+
+/**
+ * Tests {@link NumberSequenceSource}.
+ */
+public class NumberSequenceSourceITCase {
+	@Rule
+	public StreamCollector collector = new StreamCollector();
+
+	@Test
+	public void testCheckpointingWithDelayedAssignment() throws Exception {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		env.setParallelism(1);
+		env.setRestartStrategy(new RestartStrategies.NoRestartStrategyConfiguration());
+		env.enableCheckpointing(50, CheckpointingMode.EXACTLY_ONCE);
+		final SingleOutputStreamOperator<Long> stream = env
+			.fromSequence(0, 100)
+			.map(x -> {
+				Thread.sleep(10);
```

Comment:
Sleeping here for a bit longer (50 in my setup) makes the 2nd issue (missing `hasNext`) fail more likely.

Reply:
Yes, that's what I also thought. However, I also didn't want to make the test too long and was hoping that the sheer amount of AZP executions + VM hiccups actually provide enough coverage.

Reply:
Okay, I tested various sleep times and for me 50 showed the issue most reliable (10/10), so I'm going with that.

Reply:
Ah, I just noticed that it's much more likely if it's in sync with checkpointing time - so I decreased both to 10 now.